sup3r/pipeline/slicer.py

```diff
@@ -494,11 +520,50 @@ def get_padded_slices(slices, shape, enhancement, padding, step=None):
         pad = step * padding * enhancement
         pad_slices = []
         for _, s in enumerate(slices):
-            start = np.max([0, s.start * enhancement - pad])
             end = np.min([enhancement * shape, s.stop * enhancement + pad])
+            start = np.max([0, s.start * enhancement - pad])
             pad_slices.append(slice(start, end, step))
         return pad_slices
 
+    def check_boundary_slice(self, unpadded_slices, cropped_slices, dim):
+        """Check cropped slice at the right boundary for minimum shape.
+
+        It is possible for the forward pass chunk shape to divide the grid size
+        such that the last slice (right boundary) does not meet the minimum
+        number of elements. (Padding layers in the generator typically require
+        a minimum shape of 4). e.g. ``grid_size = (8, 8)`` with
+        ``fwp_chunk_shape = (7, 7, ...)`` results in unpadded slices with just
+        one element. If the padding is 0 or 1 these padded slices have length
+        less than 4. When this minimum shape is not met we apply extra padding
+        in :meth:`self._get_pad_width`. Cropped slices have to be adjusted to
+        account for this here."""
+
+        warn_msg = (
+            'The final spatial slice for dimension #%s is too small '
+            '(slice=slice(%s, %s), padding=%s). The start of this slice will '
+            'be reduced to try to meet the minimum slice length.'
+        )
+
+        lr_slice_start = unpadded_slices[-1].start or 0
+        lr_slice_stop = unpadded_slices[-1].stop or self.coarse_shape[dim]
+
+        # last slice adjustment
+        if 2 * self.spatial_pad + (lr_slice_stop - lr_slice_start) < 4:
+            logger.warning(
+                warn_msg,
+                dim + 1,
+                lr_slice_start,
+                lr_slice_stop,
+                self.spatial_pad,
+            )
+            warn(
```

Comment:
I feel like this should almost be an exception and force you to adjust your fwp shape... It could result in some pretty garbage data. Does this increase the chunk size for the last chunk and crop more of the output? And leave the neighboring chunk alone? Actually that makes sense. Does this check vertical chunk dims too?

Reply:
Yeah this pads only the "final" chunk and then crops the output more. This is done for each spatial dimension so it ends up padding the right-most and bottom-most boundary. True edge case here since it almost never happens when you have non-zero spatial padding and then only happens if spatial padding is 1 and the last unpadded chunk size is also 1.

Reply:
Gotcha that makes sense. good solution too!